File: euxis-gateway/src/gateway/mcp.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2024-2026 Sebastien Rousseau ᛫ https://sebastienrousseau.com

"""Model Context Protocol (MCP) Host implementation for Euxis Gateway."""

from __future__ import annotations

import asyncio
import json
import logging
import time
from typing import Any, Dict, List, Optional, Callable, Awaitable

from fastapi import WebSocket, WebSocketDisconnect

logger = logging.getLogger("euxis.gateway.mcp")

# MCP Protocol Version
MCP_VERSION = "2024-11-05"

class MCPHost:
    """MCP Host that manages tools, resources, and client connections."""

    def __init__(self, state: Any, config: Dict[str, Any]):
        self.state = state
        self.config = config
        self.tools: Dict[str, Callable[..., Awaitable[Any]]] = {}
        self.resources: Dict[str, Dict[str, Any]] = {}
        self._setup_default_tools()

    def _setup_default_tools(self):
        """Register built-in Euxis tools as MCP tools."""
        self.register_tool(
            "get_metrics",
            "Retrieve real-time metrics from the Euxis mesh.",
            {
                "type": "object",
                "properties": {
                    "session_id": {"type": "string", "description": "Optional session ID to filter metrics."}
                }
            },
            self._tool_get_metrics
        )
        self.register_tool(
            "sign_payload",
            "Sign a payload using Euxis Cryptographic Provenance.",
            {
                "type": "object",
                "properties": {
                    "payload": {"type": "string", "description": "The payload to sign."},
                    "key_id": {"type": "string", "description": "The cryptographic key ID to use."}
                },
                "required": ["payload"]
            },
            self._tool_sign_payload
        )

    def register_tool(self, name: str, description: str, input_schema: Dict[str, Any], handler: Callable[..., Awaitable[Any]]):
        """Register a new tool with the MCP host."""
        self.tools[name] = {
            "name": name,
            "description": description,
            "inputSchema": input_schema,
            "handler": handler
        }
        logger.info(f"Registered MCP tool: {name}")

    async def handle_connection(self, websocket: WebSocket):
        """Handle an incoming MCP client connection over WebSocket."""
        await websocket.accept()
        client_info = {}
        initialized = False

        try:
            while True:
                message_text = await websocket.receive_text()
                try:
                    request = json.loads(message_text)
                except json.JSONDecodeError:
                    await self._send_error(websocket, None, -32700, "Parse error")
                    continue

                req_id = request.get("id")
                method = request.get("method")
                params = request.get("params", {})

                if not initialized and method != "initialize":
                    await self._send_error(websocket, req_id, -32002, "Not initialized")
                    continue

                if method == "initialize":
                    client_info = params.get("clientInfo", {})
                    initialized = True
                    await self._send_result(websocket, req_id, {
                        "protocolVersion": MCP_VERSION,
                        "capabilities": {
                            "tools": {"listChanged": False},
                            "resources": {"subscribe": True, "listChanged": False}
                        },
                        "serverInfo": {"name": "euxis-gateway", "version": "0.1.0"}
                    })
                elif method == "notifications/initialized":
                    # Client acknowledgment of initialization
                    pass
                elif method == "tools/list":
                    tools_list = [
                        {"name": t["name"], "description": t["description"], "inputSchema": t["inputSchema"]}
                        for t in self.tools.values()
                    ]
                    await self._send_result(websocket, req_id, {"tools": tools_list})
                elif method == "tools/call":
                    tool_name = params.get("name")
                    tool_args = params.get("arguments", {})
                    if tool_name in self.tools:
                        try:
                            result = await self.tools[tool_name]["handler"](**tool_args)
                            await self._send_result(websocket, req_id, {"content": [{"type": "text", "text": json.dumps(result)}]})
                        except Exception as e:
                            await self._send_error(websocket, req_id, -32603, f"Internal error: {str(e)}")
                    else:
                        await self._send_error(websocket, req_id, -32601, f"Tool not found: {tool_name}")
                elif method == "ping":
                    await self._send_result(websocket, req_id, {})
                else:
                    await self._send_error(websocket, req_id, -32601, f"Method not found: {method}")

        except WebSocketDisconnect:
            logger.info("MCP client disconnected")
        except Exception as e:  # pragma: no cover
            logger.error(f"MCP connection error: {str(e)}")

    async def _send_result(self, websocket: WebSocket, req_id: Any, result: Any):
        """Send a JSON-RPC result response."""
        await websocket.send_text(json.dumps({
            "jsonrpc": "2.0",
            "id": req_id,
            "result": result
        }))

    async def _send_error(self, websocket: WebSocket, req_id: Any, code: int, message: str):
        """Send a JSON-RPC error response."""
        await websocket.send_text(json.dumps({
            "jsonrpc": "2.0",
            "id": req_id,
            "error": {"code": code, "message": message}
        }))

    # --- Tool Handlers ---

    async def _tool_get_metrics(self, session_id: Optional[str] = None) -> Dict[str, Any]:
        """Tool handler for retrieving Euxis metrics."""
        # Integration with STATE and euxis-metrics
        return {
            "uptime_ms": int((time.time() - self.state.started_at) * 1000),
            "sessions_active": self.state.sessions_active,
            "session_id": session_id,
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        }

    async def _tool_sign_payload(self, payload: str, key_id: Optional[str] = None) -> Dict[str, Any]:
        """Tool handler for signing payloads."""
        # Integration with euxis-crypto-lib
        # For now, a mock implementation that would call into the Rust crypto lib
        import hashlib
        signature = hashlib.sha256(payload.encode()).hexdigest()
        return {
            "payload": payload,
            "signature": f"euxis_sig_{signature}",
            "key_id": key_id or "default_gateway_key",
            "algorithm": "Ed25519"
        }

File: apps/gateway/src/mcp.cpp
#include "euxis/gateway/mcp.hpp"
#include "euxis/gateway/state.hpp"
#include "euxis/crypto/ed25519.hpp"

#include <chrono>
#include <spdlog/spdlog.h>
#include <sstream>
#include <iomanip>

namespace euxis::gateway {

namespace {

auto hex_encode(std::span<const std::byte> data) -> std::string {
    std::ostringstream oss;
    for (auto b : data) {
        oss << std::hex << std::setw(2) << std::setfill('0')
            << static_cast<int>(b);
    }
    return oss.str();
}

} // namespace

McpHost::McpHost(const nlohmann::json& config) : config_(config) {
    setup_default_tools();
}

void McpHost::register_tool(std::string name, std::string description,
                             nlohmann::json input_schema,
                             McpToolHandler handler) {
    tools_.push_back(
        McpTool{.name = name,
                .description = std::move(description),
                .input_schema = std::move(input_schema)});
    handlers_[std::move(name)] = std::move(handler);
}

void McpHost::register_resource(const std::string& uri,
                                 const std::string& name,
                                 const std::string& description,
                                 const std::string& mime_type,
                                 McpResourceHandler handler) {
    resources_.push_back(McpResource{
        .uri = uri,
        .name = name,
        .description = description,
        .mime_type = mime_type,
    });
    resource_handlers_[uri] = std::move(handler);
}

auto McpHost::handle_request(const nlohmann::json& request) -> nlohmann::json {
    if (!request.contains("jsonrpc") || request["jsonrpc"] != "2.0" ||
        !request.contains("method")) {
        return make_error(request.value("id", nlohmann::json(nullptr)),
                          -32600, "Invalid Request");
    }

    const auto& method = request["method"].get<std::string>();
    auto id = request.value("id", nlohmann::json(nullptr));

    // initialize is always allowed
    if (method == "initialize") {
        return handle_initialize(request);
    }

    // notifications/initialized is a no-op ack
    if (method == "notifications/initialized") {
        return make_result(id, nlohmann::json::object());
    }

    // Guard: reject non-initialize methods before initialization
    if (!initialized_) {
        return make_error(id, -32002, "Server not initialized");
    }

    if (method == "tools/list") {
        return handle_tools_list(request);
    }
    if (method == "tools/call") {
        return handle_tools_call(request);
    }
    if (method == "resources/list") {
        return handle_resources_list(request);
    }
    if (method == "resources/read") {
        return handle_resources_read(request);
    }
    if (method == "ping") {
        return handle_ping(request);
    }

    return make_error(id, -32601, "Method not found");
}

auto McpHost::tools() const -> const std::vector<McpTool>& {
    return tools_;
}

auto McpHost::resources() const -> const std::vector<McpResource>& {
    return resources_;
}

auto McpHost::handle_initialize(const nlohmann::json& req) -> nlohmann::json {
    initialized_ = true;
    nlohmann::json result = {
        {"protocolVersion", protocol_version_},
        {"capabilities", {{"tools", {{"listChanged", true}}},
                          {"resources", {{"listChanged", true}}}}},
        {"serverInfo", {{"name", "euxis-gateway"}, {"version", "0.0.3"}}},
    };
    return make_result(req.value("id", nlohmann::json(nullptr)), result);
}

auto McpHost::handle_tools_list(const nlohmann::json& req) -> nlohmann::json {
    auto tools_json = nlohmann::json::array();
    for (const auto& tool : tools_) {
        tools_json.push_back({
            {"name", tool.name},
            {"description", tool.description},
            {"inputSchema", tool.input_schema},
        });
    }
    return make_result(req.value("id", nlohmann::json(nullptr)),
                       {{"tools", tools_json}});
}

auto McpHost::handle_tools_call(const nlohmann::json& req) -> nlohmann::json {
    auto id = req.value("id", nlohmann::json(nullptr));
    auto params = req.value("params", nlohmann::json::object());
    auto tool_name = params.value("name", std::string{});

    auto it = handlers_.find(tool_name);
    if (it == handlers_.end()) {
        return make_error(id, -32601,
                          "Unknown tool: " + tool_name);
    }

    try {
        auto args = params.value("arguments", nlohmann::json::object());
        auto result = it->second(args);
        nlohmann::json content = {{{"type", "text"},
                                   {"text", result.dump()}}};
        return make_result(id, {{"content", content}});
    } catch (const std::exception& e) {
        return make_error(id, -32603, std::string("Tool error: ") + e.what());
    }
}

auto McpHost::handle_resources_list(const nlohmann::json& req)
    -> nlohmann::json {
    auto resources_json = nlohmann::json::array();
    for (const auto& r : resources_) {
        resources_json.push_back({
            {"uri", r.uri},
            {"name", r.name},
            {"description", r.description},
            {"mimeType", r.mime_type},
        });
    }
    return make_result(req.value("id", nlohmann::json(nullptr)),
                       {{"resources", resources_json}});
}

auto McpHost::handle_resources_read(const nlohmann::json& req)
    -> nlohmann::json {
    auto id = req.value("id", nlohmann::json(nullptr));
    auto params = req.value("params", nlohmann::json::object());
    auto uri = params.value("uri", std::string{});

    auto it = resource_handlers_.find(uri);
    if (it == resource_handlers_.end()) {
        return make_error(id, -32602, "Unknown resource URI: " + uri);
    }

    try {
        auto content = it->second(uri);
        return make_result(id, {{"contents", {{
            {"uri", uri},
            {"mimeType", "application/json"},
            {"text", content.dump()},
        }}}});
    } catch (const std::exception& e) {
        return make_error(id, -32603,
                          std::string("Resource error: ") + e.what());
    }
}

auto McpHost::handle_ping(const nlohmann::json& req) -> nlohmann::json {
    return make_result(req.value("id", nlohmann::json(nullptr)),
                       nlohmann::json::object());
}

auto McpHost::make_result(const nlohmann::json& id,
                           const nlohmann::json& result) -> nlohmann::json {
    return {{"jsonrpc", "2.0"}, {"id", id}, {"result", result}};
}

auto McpHost::make_error(const nlohmann::json& id, int code,
                          const std::string& msg) -> nlohmann::json {
    return {{"jsonrpc", "2.0"},
            {"id", id},
            {"error", {{"code", code}, {"message", msg}}}};
}

void McpHost::setup_default_tools() {
    register_tool(
        "get_metrics", "Return current gateway metrics",
        {{"type", "object"}, {"properties", nlohmann::json::object()}},
        [](const nlohmann::json& /*args*/) -> nlohmann::json {
            auto now = std::chrono::steady_clock::now();
            auto uptime = std::chrono::duration_cast<std::chrono::milliseconds>(
                              now.time_since_epoch())
                              .count();
            return {{"uptime_ms", uptime},
                    {"sessions_active", 0},
                    {"timestamp", timestamp()}};
        });

    register_tool(
        "sign_payload", "Sign a payload with Ed25519",
        {{"type", "object"},
         {"properties",
          {{"payload", {{"type", "string"}, {"description", "Data to sign"}}}}},
         {"required", {"payload"}}},
        [](const nlohmann::json& args) -> nlohmann::json {
            auto payload = args.value("payload", std::string{});
            auto kp = euxis::crypto::generate_keypair();
            std::vector<std::byte> msg(payload.size());
            std::memcpy(msg.data(), payload.data(), payload.size());
            auto sig = euxis::crypto::sign(
                std::span<const std::byte, 64>(kp.secret_key), msg);
            return {{"payload", payload},
                    {"signature", hex_encode(sig)},
                    {"algorithm", "ed25519"}};
        });
}

} // namespace euxis::gateway
